cli/azd/pkg/tools/bicep/bicep.go

```diff
@@ -5,170 +5,225 @@ package bicep
+	transporter policy.Transporter,
+) (BicepCli, error) {
+	if override := os.Getenv("AZD_BICEP_TOOL_PATH"); override != "" {
+		log.Printf("using external bicep tool: %s", override)
+
+		return &bicepCli{
+			path:   override,
+			runner: commandRunner,
+		}, nil
+	}
 
-func (cli *bicepCli) Name() string {
-	return "Bicep CLI"
-}
+	bicepPath, err := cachedBicepPath()
+	if err != nil {
+		return nil, fmt.Errorf("finding bicep: %w", err)
+	}
 
-func (cli *bicepCli) InstallUrl() string {
-	return "https://aka.ms/azure-dev/bicep-install"
-}
+	if _, err = os.Stat(bicepPath); err != nil && !errors.Is(err, os.ErrNotExist) {
+		return nil, fmt.Errorf("finding bicep: %w", err)
+	} else if errors.Is(err, os.ErrNotExist) {
+		if err := os.MkdirAll(filepath.Dir(bicepPath), osutil.PermissionDirectory); err != nil {
+			return nil, fmt.Errorf("downloading bicep: %w", err)
+		}
 
-func (cli *bicepCli) versionInfo() tools.VersionInfo {
-	return tools.VersionInfo{
-		MinimumVersion: semver.Version{
-			Major: 0,
-			Minor: 8,
-			Patch: 9,
-		},
-		//nolint:lll
-		UpdateCommand: `Run 'az bicep upgrade or Visit https://aka.ms/azure-dev/bicep-install to upgrade`,
+		if err := runStep(
+			ctx, console, "Downloading Bicep", func() error {
+				return downloadBicep(ctx, transporter, cBicepVersion, bicepPath)
+			},
+		); err != nil {
+			return nil, fmt.Errorf("downloading bicep: %w", err)
+		}
 	}
-}
 
-func (cli *bicepCli) CheckInstalled(ctx context.Context) (bool, error) {
-	bicepRes, err := cli.runCommand(ctx, "--version")
-	if err != nil {
-		return false, fmt.Errorf("checking %s version: %w", cli.Name(), err)
+	cli := &bicepCli{
+		path:   bicepPath,
+		runner: commandRunner,
 	}
 
-	bicepSemver, err := tools.ExtractVersion(bicepRes.Stdout)
+	ver, err := cli.version(ctx)
 	if err != nil {
-		return false, fmt.Errorf("converting to semver version fails: %w", err)
+		return nil, fmt.Errorf("checking bicep version: %w", err)
 	}
 
-	updateDetail := cli.versionInfo()
-	if bicepSemver.LT(updateDetail.MinimumVersion) {
-		return false, &tools.ErrSemver{ToolName: cli.Name(), VersionInfo: updateDetail}
+	if ver.LT(cBicepVersion) {
+		log.Printf("installed bicep version %s is older than %s; updating.", ver.String(), cBicepVersion.String())
+
+		if err := runStep(
+			ctx, console, "Upgrading Bicep", func() error {
+				return downloadBicep(ctx, transporter, cBicepVersion, bicepPath)
+			},
+		); err != nil {
+			return nil, fmt.Errorf("upgrading bicep: %w", err)
+		}
 	}
 
-	return true, nil
+	log.Printf("using local bicep: %s", bicepPath)
+
+	return cli, nil
 }
 
-func (cli *bicepCli) Build(ctx context.Context, file string) (string, error) {
-	args := []string{"build", file, "--stdout"}
-	buildRes, err := cli.runCommand(ctx, args...)
+// runStep runs a long running operation, using the console to show a spinner for progress and status.
+func runStep(ctx context.Context, console input.Console, title string, action func() error) error {
```

Comment:
@vhvb1989 As you are thinking about UX stuff - I was unsure if there was a better way to accomplish what I wanted here.  It didn't look like the console abstraction had anything here, but it seems like "run this function and have the errorness play into the spinner" is something we'd want to build an abstraction around at some point?

I'd love long term if there was some way for me to also test that the spinner was shown (and the result of the step).  It wasn't obvious to me about how to include such assertions in my test.

I'm not blocked or anything - just things that came to mind while doing this.

Reply:
We have one API in the `spin` package (w/o UX output) to wrap a with-spinner operation:
  `func (s *Spinner) Run(runFn func() error) error {`

We also have this method `getStepResultFormat()` (currently as part of cmd package, but moved it to `output` package on my latest UX PR) which is used to get the UX element depending on an error.

However, I am trying to stay away from the pattern of wrapping a callback inside a `start()` an `stop()` spinner (like the `spin.Run()` approach), becuase I have learned that eventually, any callback might require user interaction, and the pattern needs to evolve to using an `asyncContext` (like the pattern on deploy), which I think gets things more complicated than they should.

Instead, I would like to maintain a pattern where we don't wrap the callback within a function to start(), stop() the spinner. Instead, the callback should use the console as an input and take control of start()/stop()/update()/interact() as it needs and whenever it needs to. Going from:

```go
runStep(ctx, console, title, action)
```
to:
```go
action(ctx, console)
```

The action is then free to start by calling `console.StartSpinner()` and to update the message in the spinner on demmand (no asyncContext).

Reply:
> I'd love long term if there was some way for me to also test that the spinner was shown

I've been thinking about this as well.  It is not trivial becuase sometimes we want to show the spinner and then remove it (so it won't be part of the final output).

One option I have in my mind is to use the mockedConsole to capture the output on every change, creating a snapshot /recording that we can use later to compare. However, we would be testing the mock console implementation and not the actual console implementation from the app :( .  
Another alternative is to use the `io.Writer` , changing only the writter from the defaut console, but it complicates setting the Reader as well.
The last option is to let the mockConsole to wrap the default console to use the same implementation for writing as the default but with some more code to create the recordings/snapshots

Reply:
> One option I have in my mind is to use the mockedConsole to capture the output on every change, creating a snapshot /recording that we can use later to compare.

I ended up doing something like this.  The MockConsole now logs the calls to ShowSpinner/StopSpinner and you can query that log.

So I can write a test like this (to ensure that we showed the spinner saying we were downloading Bicep and then that it succeded):

```go
	require.Equal(t, 2, len(mockContext.Console.SpinnerOps()))

	require.Equal(t, console.SpinnerOp{
		Op:      console.SpinnerOpShow,
		Message: "Upgrading Bicep",
		Format:  input.Step,
	}, mockContext.Console.SpinnerOps()[0])

	require.Equal(t, console.SpinnerOp{
		Op:      console.SpinnerOpStop,
		Message: "Upgrading Bicep",
		Format:  input.StepDone,
	}, mockContext.Console.SpinnerOps()[1])
```

Reply:
It doesn't seem correct to me that the bicepCli object should be concerned about the presentation of the bicep download.

Reply:
> It doesn't seem correct to me that the bicepCli object should be concerned about the presentation of the bicep download.

One small note (perhaps it is just semantics, but it feels somewhat meaningful here) is that the presentation of the bicep download is done by the function that constructs the bicepCli object, not the object itself.  All of this work happens in order to get an instance of the `BicepCli` interface.

We could imagine an alternate design, where NewBicepCli looks like this:

```go
func NewBicepCli(bicepPath string, commandRunner exec.CommandRunner) (BicepCli, error) {
    // stat path and ensure it exists and return the implementation object
}
```

And we move the logic that is concerned with downloading the copy of bicep (basically the implementation of `newBicepCliWithTransporter` and the transitive closure of stuff it touches, which ends up being most of `bicep.go`) into `bicep_provider.go` in `pkg/infra/provisioning/bicep`

Maybe this feels better because that logic is in the bicep provider package instead of the bicep cli tool package?

Happy to make the change - I can see the argument for making the change (that the policy over where the tool is cached, and the like is more tied to the bicep provider than the tool itself)

@wbreza Would that address your concerns or does that factoring still feel incorrect?

Reply:
Going to merge this as is - but let me know if you'd prefer this refactoring, @wbreza and I will push it through.